### ISSUE TITLE
Drag and Drop simulation does not trigger dragstart+drop events but does mousedown+mouseup

### ISSUE THREAD

**Issue by isaiahgrant** [laravel/dusk#308]
**NOTE:** This is not an issue, but rather a temporary alternative for developers (like myself) whom are looking for their tests to function while this issue remains open. I would have liked to comment on the existing issue, but it has been closed for comments (:-/).

## Meta -
OS:  Ubuntu 16.04
Selenium Version:  PHPUnit_Selenium 5.0.8 - 2015-10-23
Browser:  Chrome

Browser Version:  Version 58.0.3029.110 Built on Ubuntu , running on Ubuntu 16.04 (64-bit)

## Expected Behavior -
As already known, drag and drop events using draggable and droppable attributes/functionality does not function as expected when ran in Laravel's Dusk tests (among others). Running a $browser->drag('#from','#to') from testers should automate dragging the "from" element to the "to" element firing the "dragstart", "dragover", and "drop" events. 

## Actual Behavior -
```
/**
     * Drag and drop from $source to $target.
     *
     * @param WebDriverElement $source
     * @param WebDriverElement $target
     * @return WebDriverActions
     */
    public function dragAndDrop(WebDriverElement $source, WebDriverElement $target)
    {
        $this->action->addAction(
            new WebDriverClickAndHoldAction($this->mouse, $source)
        );
        $this->action->addAction(
            new WebDriverMouseMoveAction($this->mouse, $target)
        );
        $this->action->addAction(
            new WebDriverButtonReleaseAction($this->mouse, $target)
        );

        return $this;
    }
```

This behavior, though in theory simulates a drag and drop, actually fires "mousedown" and "mouseup" events rather than the "dragstart", "dragover", and "drop" events.

## Steps to reproduce -
```
$this->browse(function (Browser $browser) {
    $browser->setUp()->visit($this->url)->drag('#element-a', '#element-b');
}
```

## **Temporary solution**
Create events for "mousedown" and "mouseup", on draggable and droppable elements (respectively), that trigger the dragstart and drop events (respectively).

 -- Example:
Included is a jsfiddle link that provides the code necessary for user driven drag-and-drop as well as the Dusk test drag and drop simulation (though of course the dusk test cannot be ran from it):

[Dusk "testable" Drag and Drop Fiddle](https://jsfiddle.net/grantdscs/nehjrbzo/)

A highlight of the functionality:
```
    $("a[draggable='true']").on("mousedown", function(event) {
        $(this).trigger("dragstart");
    });

    $("div[droppable='true']").on("mouseup", function(event) {
        $(this).trigger("drop");
    });
```

**Comment by nbyloff** [comment]
I am having this issue and found this issue on Selenium: https://github.com/SeleniumHQ/selenium/issues/3269

Although it doesn't currently work in Dusk, it is a working solution for other languages. I just haven't figured out how to modify the dragAndDrop Dusk method yet. I tried this along the lines of what's suggested in that post, still working through it but wanted to share in case someone who is more familiar with Dusk to take a stab at it.

```
    public function dragAndDrop(WebDriverElement $source, WebDriverElement $target)
    {
        $this->action->addAction(
            new WebDriverClickAndHoldAction($this->mouse, $source)
        );

	// https://github.com/SeleniumHQ/selenium/issues/3269
	$this->action->addAction(
		new WebDriverMoveToOffsetAction($this->mouse, $source, -1, -1)
	);

        $this->action->addAction(
            new WebDriverMouseMoveAction($this->mouse, $target)
        );
        $this->action->addAction(
            new WebDriverButtonReleaseAction($this->mouse, $target)
        );

        return $this;
    }))

```

**Comment by lk77** [comment]
Hello,

the best way to do this, is by cheating :

```
$browser->script("document.querySelector('.draggable-list').__vue__.updatePosition($fromIndex, $toIndex)");
```

**Comment by rvalim** [comment]
Hello Guys, 

Is there same prediction for solution??

Tks

**Comment by alainmoran** [comment]
So after tearing my hair out over this issue for a couple of days, I finally realised what *my* problem was!

If you've been reading all the other issues relating to this problem then you'll see a lot of people suggesting splitting the dragAndDrop down into its action components ... which didn't work for me unless I dropped in a breakpoint and stepped through the process.

I tried adding pauses and that, but nothing worked until I finally realised that the problem isn't with the dragging and dropping, it's that the test isn't waiting for the js (in my case ts) code to actually pick the item up and start dragging it.

Your case will probably be different (I'm using https://www.npmjs.com/package/react-list-drag-and-drop), but it's possible you may be able to adapt my solution for your use case:

```
    private void waitUntilAnyVisible(final By locator, final SearchContext container, int seconds) {
        try {
            // wait for any elements matching the locator to be visible
            waitModel().withTimeout(seconds, TimeUnit.SECONDS).until(new Predicate<WebDriver>() {
                @Override
                public boolean apply(WebDriver input) {
                    return ((container!=null) ? container : input).findElements(locator).stream()
                            .anyMatch(WebElement::isDisplayed);
                }
            });
        } catch (IndexOutOfBoundsException | TimeoutException e) {
            // do nothing - it's not there!
        }
    }

    public void dragDrop(WebElement sourceElement, WebElement destinationElement) {
        new Actions(browser).clickAndHold(sourceElement).moveByOffset(-10,0).perform();
        waitUntilAnyVisible(By.cssSelector(".dragged"), root, 10);
        waitUntilAnyVisible(By.cssSelector(".floating"), root, 10);
        new Actions(browser).moveToElement(destinationElement).release().perform();
    }
```

**Comment by lk77** [comment]
Hello,

i'm having this issue again, 
but instead of cheating by interacting with the draggable in js,
i was able to make it work by enabling `:force-fallback:"forceFallback"` option

```
if(navigator.userAgent !== 'laravel/dusk') {
   this.forceFallback =true;
}
```

**Comment by ziadoz** [comment]
If you're testing Sortable JS, you can create a plugin that forces the fallback mode mentioned by @lk77 everywhere:

Add a `resources/js/testing.js` script, and then add it to the `vite.config.js` manifest: 

```
import Sortable from 'sortablejs';

Sortable.mount(function (sortable) {
    sortable.options.forceFallback = true;
});
```

Then use it conditionally inside your Blade layout template:
```
@if (app()->runningUnitTests())
    @vite('resources/js/testing.js')
@endif
```

**Comment by lk77** [comment]
@ziadoz `app()->runningUnitTests()` will be false inside dusk tests however

**Comment by ziadoz** [comment]
> @ziadoz `app()->runningUnitTests()` will be false inside dusk tests however

@lk77 Ah yes, good point. I run my Dusk tests with `--env=testing`, which would explain why it works.

**Comment by lk77** [comment]
@ziadoz even that would not work, 
when i say inside dusk tests, i mean inside the browser

**Comment by ziadoz** [comment]
> the only way to make it work is to change the APP_ENV in the .env file to testing, because that's the one laravel is gonna use when receiving a request from the headless browser

Yep, this is how my app is configured. Two separate domains `local.app.com` and `testing.app.com`, both with the correct `APP_ENV` set in their Apache configurations.

Anyway, the bit I wanted to share is using a Sortable JS plugin to change the fallback mode on every sortable in one place. It seemed like it might be helpful to someone else.

**Comment by ziadoz** [comment]
One last thing I'd add here is that you can set a custom user agent using a Chrome flag: `'--user-agent=Laravel/Dusk'` 

Using the experimental option approach seems to trigger a legacy emulation mode:

```
[1701448260.274][INFO]: Operating in legacy emulation mode as 'mobileEmulation' contains no 'clientHints'.
[1701448260.274][INFO]: No 'clientHints' found. Operating in legacy mode. Inferring clientHints as: {architecture='', bitness='', brands=<browser-defined>, fullVersionList=<browser-defined>, mobile=false, model='', platform='', platformVersion='', wow64=false}
[1701448260.275][INFO]: Operating in legacy emulation mode as 'mobileEmulation' contains no 'clientHints'.
[1701448260.275][INFO]: No 'clientHints' found. Operating in legacy mode. Inferring clientHints as: {architecture='', bitness='', brands=<browser-defined>, fullVersionList=<browser-defined>, mobile=false, model='', platform='', platformVersion='', wow64=false}
```
